services/datamanager/builtin/capture_test.go

```diff
@@ -235,13 +198,126 @@ func TestDataCaptureEnabled(t *testing.T) {
 	}
 }
 
-func waitForCaptureFiles(captureDir string) {
+func TestSwitchResource(t *testing.T) {
+	captureDir := t.TempDir()
+	mockClock := clk.NewMock()
+	// Make mockClock the package level clock used by the dmsvc so that we can simulate time's passage
+	clock = mockClock
+
+	// Set up robot config.
+	config, deps := setupConfig(t, enabledTabularCollectorConfigPath)
+	config.CaptureDisabled = false
+	config.ScheduledSyncDisabled = true
+	config.CaptureDir = captureDir
+
+	// Build and start data manager.
+	dmsvc, r := newTestDataManager(t)
+	defer func() {
+		test.That(t, dmsvc.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	resources := resourcesFromDeps(t, r, deps)
+	err := dmsvc.Reconfigure(context.Background(), resources, resource.Config{
+		ConvertedAttributes: config,
+	})
+	test.That(t, err, test.ShouldBeNil)
+	passTimeCtx1, cancelPassTime1 := context.WithCancel(context.Background())
+	donePassingTime1 := passTime(passTimeCtx1, mockClock, captureInterval)
+
+	waitForCaptureFilesToExceedN(captureDir, 0)
+	testFilesContainSensorData(t, captureDir, nil)
+
+	cancelPassTime1()
+	<-donePassingTime1
+
+	// Change the resource named arm1 to show that the data manager recognizes that the collector has changed with no other config changes.
+	for resource := range resources {
+		if resource.Name == "arm1" {
+			newResource := inject.NewArm(resource.Name)
+			newResource.EndPositionFunc = func(ctx context.Context, extra map[string]interface{}) (spatialmath.Pose, error) {
+				// Return a different value from the initial arm1 resource.
+				return spatialmath.NewPoseFromPoint(r3.Vector{X: 888, Y: 888, Z: 888}), nil
+			}
+			resources[resource] = newResource
+		}
+	}
+
+	err = dmsvc.Reconfigure(context.Background(), resources, resource.Config{
+		ConvertedAttributes: config,
+	})
+	test.That(t, err, test.ShouldBeNil)
+
+	passTimeCtx2, cancelPassTime2 := context.WithCancel(context.Background())
+	donePassingTime2 := passTime(passTimeCtx2, mockClock, captureInterval)
+
+	// Test that sensor data is captured from the new collector.
+	waitForCaptureFilesToExceedN(captureDir, len(getAllFileInfos(captureDir)))
+	testFilesContainSensorData(t, captureDir, func(t *testing.T, sd []*v1.SensorData) {
+		valueSet := map[float64]struct{}{}
+		for _, d := range sd {
+			valueSet[d.GetStruct().GetFields()["Number"].GetStructValue().GetFields()["Dual"].GetStructValue().GetFields()["Jmag"].GetNumberValue()] = struct{}{}
+		}
+		// Each resource's mocked capture method outputs a different value. Assert that we see data captured by the initial arm1 resource as well as the changed resource.
```

Comment:
For:
> Assert that we see data captured by the initial arm1 resource as well as the changed resource.

I feel like this is a little ambiguous w.r.t. if we expect the old collector to still be capturing data - we don't, right? Should there be an assertion ensuring that, too? I think this would still pass even if the old collector was still running

Reply:
Yup, correct that we don't want the old collector to still be capturing data. That assertion would be super easy to write if we were using a different captureDir for each resource (we could just assert that the initial captureDir has the same amount of files after the second passTime), but unfortunately that would cause a config change.

One way I can test this is to assert that the number of datapoints that correlate to the old collector doesn't increase after switching out the resource - this will be very dependent on the test implementation that each collector will always return the same value (like the check that data is collected from the new collector), so happy to hear any thoughts you might have on making these assertions more generic.

Reply:
> but unfortunately that would cause a config change.

Why do we want to avoid that?

Reply:
The goal of this test is to make sure that the collector reinits when there is a resource change but all other fields in the DataCaptureConfig stays the same: ticket [here](https://viam.atlassian.net/browse/RSDK-2929)

Reply:
I see, thanks!